Cldc 1671 deactivate scheme

natdeanlewissoftwire opened the pull request:
ticket: https://digital.dclg.gov.uk/jira/browse/CLDC-1671
tested locally
risk of conflict with simultaneous location deactivation work by @kosiakkatrina but we've been communicating to ameliorate this
no docs required

This is quite a large ticket, so this PR doesn't include all the functionality yet - updating logs will be done separately



deactivation:
<img width="727" alt="image" src="https://user-images.githubusercontent.com/94526761/201646617-89099ef7-1686-4402-8f4a-637a02c92050.png">

confirmation:
<img width="949" alt="image" src="https://user-images.githubusercontent.com/94526761/201646639-f0fefcb1-75d2-4c4c-8940-d7cbe3ea66d6.png">

success: 
<img width="976" alt="image" src="https://user-images.githubusercontent.com/94526761/201646572-295291b0-a673-4062-b116-a5c2885f0128.png">

natdeanlewissoftwire committed:
feat: wip update scheme summary page

natdeanlewissoftwire committed:
feat: wip deactivate scheme schemes page

natdeanlewissoftwire committed:
feat: wip toggle active page

kosiakkatrina commented on app/views/schemes/show.html.erb:
```diff
@@ -15,7 +15,7 @@
 
 <%= govuk_summary_list do |summary_list| %>
-  <% @scheme.display_attributes.each do |attr| %>
+  <% display_attributes(@scheme).each do |attr| %>
```
Nice!

natdeanlewissoftwire replied:
Thanks - stole this from your PR 👀

kosiakkatrina replied:
I should've updated it really 👀

kosiakkatrina committed:
Change conditional question controller to accommodate all models

natdeanlewissoftwire committed:
feat: wip set deactivation_date to a datetime (to be more specific times later_

natdeanlewissoftwire committed:
feat: add specific datetimes for deactivation

natdeanlewissoftwire committed:
feat: correct date and add notice

natdeanlewissoftwire committed:
feat: wip error behaviour

natdeanlewissoftwire committed:
feat: wip errors

natdeanlewissoftwire committed:
feat: wip errors refactoring

natdeanlewissoftwire committed:
feat: wip errors more refactoring

natdeanlewissoftwire committed:
refactor: linting

natdeanlewissoftwire committed:
feat: add second error in correct position and wip date range error

natdeanlewissoftwire committed:
feat: remove unneccessary db field

natdeanlewissoftwire committed:
feat: change type values to strings

natdeanlewissoftwire committed:
refactor: tidy up controller logic

natdeanlewissoftwire committed:
refactor: use same structure as locations deactivation

natdeanlewissoftwire committed:
feat: add general partially missing date input error

natdeanlewissoftwire committed:
Merge branch 'main' into CLDC-1671-deactivate-scheme

# Conflicts:
#	app/helpers/tag_helper.rb
#	config/initializers/feature_toggle.rb
#	config/locales/en.yml
#	db/schema.rb

natdeanlewissoftwire committed:
feat: add tests ("updates existing scheme with valid deactivation date and renders scheme page" still wip) and add new partially nil date behaviour to locations controller

natdeanlewissoftwire committed:
feat: fix tests, add status tag behaviour and remove unnecessary nils

natdeanlewissoftwire committed:
refactor: linting

natdeanlewissoftwire committed:
refactor: erblinting

jamesrose commented on app/controllers/schemes_controller.rb:
```diff
@@ -251,12 +272,55 @@ def find_resource
+      elsif !Date.new(year.to_i, month.to_i, day.to_i).between?(collection_start_date, Date.new(2200, 1, 1))
+        @scheme.errors.add(:deactivation_date, message: I18n.t("validations.scheme.deactivation_date.out_of_range", date: collection_start_date.to_formatted_s(:govuk_date)))
+      end
+    end
```
These validations are likely better suited to the model. Was there a reason for having them here?

natdeanlewissoftwire replied:
Nope, I'll move them across 👍

jamesrose commented on app/helpers/schemes_helper.rb:
```diff
@@ -0,0 +1,34 @@
+    base_attributes
+  end
+
+  def availability_text(scheme)
```
I don't see this method used. Is it?

natdeanlewissoftwire replied:
Ah thanks good spot, that's now redundant

jamesrose commented on app/views/schemes/show.html.erb:
```diff
@@ -15,12 +15,20 @@
+    <%= govuk_button_link_to "Deactivate this scheme", scheme_deactivate_path(@scheme), warning: true %>
+  <% else %>
+    <%= govuk_button_link_to "Reactivate this scheme", scheme_reactivate_path(@scheme) %>
+  <% end %><% end %>
```
```suggestion
<% if FeatureToggle.scheme_toggle_enabled? %>
  <% if @scheme.active? %>
    <%= govuk_button_link_to "Deactivate this scheme", scheme_deactivate_path(@scheme), warning: true %>
  <% else %>
    <%= govuk_button_link_to "Reactivate this scheme", scheme_reactivate_path(@scheme) %>
  <% end %>
<% end %>
```

Shall we add a helper method for this?

jamesrose commented on app/views/schemes/toggle_active_confirm.html.erb:
```diff
@@ -0,0 +1,18 @@
+  <% end %>
+  <h1 class="govuk-heading-l">
+    <span class="govuk-caption-l"><%= @scheme.service_name %></span>
+    <%= "This change will affect #{@scheme.lettings_logs.count} logs" %>
```
```suggestion
    This change will affect <%= @scheme.lettings_logs.count %> logs
```

jamesrose commented on app/views/schemes/toggle_active.html.erb:
```diff
@@ -0,0 +1,35 @@
+                                 **basic_conditional_html_attributes({ "deactivation_date" => %w[other] }, "scheme") do %>
+          <%= f.govuk_date_field :deactivation_date,
+                                 legend: { text: "Date", size: "m" },
+                                 hint: { text: "For example, 27 3 2008" },
```
```suggestion
                                 hint: { text: "For example, 27 3 2022" },
```

This decade 😄

natdeanlewissoftwire committed:
refactor: remove redundant line

natdeanlewissoftwire committed:
refactor: respond to PR comments 1

natdeanlewissoftwire committed:
refactor: respond to PR comments 2

natdeanlewissoftwire committed:
refactor: respond to PR comments 3

natdeanlewissoftwire committed:
refactor: respond to PR comments 3 (locations side)

natdeanlewissoftwire committed:
fix: remove @locations in location model

natdeanlewissoftwire committed:
fix: remove @locations in location model

natdeanlewissoftwire committed:
fix: update status names

natdeanlewissoftwire committed:
feat: wip validation update

natdeanlewissoftwire committed:
feat: add validation to model layer

natdeanlewissoftwire committed:
feat: further separate scheme deactivation behaviour

natdeanlewissoftwire committed:
test: update tests to new flow

natdeanlewissoftwire commented on app/models/scheme.rb:
```diff
@@ -230,4 +209,40 @@ def validate_confirmed
+    return :active if deactivation_date.blank?
+    return :deactivating_soon if Time.zone.now < deactivation_date
+    return :deactivated if Time.zone.now >= deactivation_date
+  end
```
Cheers :)

jamesrose commented on app/controllers/schemes_controller.rb:
```diff
@@ -21,6 +21,32 @@ def show
+  end
+
+  def deactivate
+    prepare_for_validation
```
No need to extract these — it's clear if everything is in one place.

jamesrose commented on app/controllers/schemes_controller.rb:
```diff
@@ -251,12 +278,40 @@ def find_resource
+    return if params[:scheme].blank?
+
+    collection_start_date = FormHandler.instance.current_collection_start_date
+    return collection_start_date if params[:scheme][:deactivation_date_type] == "default"
```
```suggestion
    if params[:scheme][:deactivation_date_type] == "default"
      return FormHandler.instance.current_collection_start_date
    end
```

natdeanlewissoftwire committed:
feat: respond to pr comments and add dynamic success text

natdeanlewissoftwire committed:
feat: duplicate behaviour schemes -> locations (+ tests)

natdeanlewissoftwire committed:
refactor: linting

natdeanlewissoftwire committed:
refactor: typo and remove unnecessary line for this PR

natdeanlewissoftwire committed:
refactor: feature toggle simplification

kosiakkatrina commented on app/models/scheme.rb:
```diff
@@ -230,4 +210,36 @@ def validate_confirmed
+  end
+
+  def implicit_run_deactivation_validations
+    deactivation_date.present? || @run_deactivation_validations
```
Would this work instead? Wouldn't need to pass run_deactivation_validations parameter if so

`deactivation_date.present? || :deactivation_date_type_changed?`

natdeanlewissoftwire replied:
I'm not sure what that _changed? method is - do you have a link to some docs?

jamesrose commented on app/controllers/locations_controller.rb:
```diff
@@ -164,47 +174,35 @@ def valid_location_admin_district?(location_params)
+    return nil if [day, month, year].any?(&:blank?)
 
-    Date.new(year.to_i, month.to_i, day.to_i)
+    Time.utc(year.to_i, month.to_i, day.to_i) if Date.valid_date?(year.to_i, month.to_i, day.to_i)
```
```suggestion
    Time.local(year.to_i, month.to_i, day.to_i) if Date.valid_date?(year.to_i, month.to_i, day.to_i)
```

Always use local times. Rails does the conversion automatically.

jamesrose commented on app/controllers/locations_controller.rb:
```diff
@@ -164,47 +174,35 @@ def valid_location_admin_district?(location_params)
     day = params[:location]["deactivation_date(3i)"]
     month = params[:location]["deactivation_date(2i)"]
     year = params[:location]["deactivation_date(1i)"]
+    return nil if [day, month, year].any?(&:blank?)
```
```suggestion
    return if [day, month, year].any?(&:blank?)
```

jamesrose commented on app/controllers/locations_controller.rb:
```diff
@@ -164,47 +174,35 @@ def valid_location_admin_district?(location_params)
   def confirm_deactivation
-    if @location.update(deactivation_date: params[:location][:deactivation_date])
-      flash[:notice] = "#{@location.name || @location.postcode} has been deactivated"
+    if @location.update!(deactivation_date: @location.deactivation_date)
```
IMO I don't see much benefit on extracting these methods.

natdeanlewissoftwire committed:
Merge branch 'main' into CLDC-1671-deactivate-scheme

# Conflicts:
#	db/schema.rb

jamesrose committed:
Refactor locations and schemes controller actions

- Rename confirmation partials to `deactivate_confirm.html.erb` so that they match the actions in which they belong to
- Make all deactivation date comparision UTC time

natdeanlewissoftwire committed:
Merge remote-tracking branch 'origin/CLDC-1671-deactivate-scheme' into CLDC-1671-deactivate-scheme

# Conflicts:
#	db/schema.rb

natdeanlewissoftwire committed:
feat: update deactivation_date validation and add tests

natdeanlewissoftwire committed:
refactor: linting

natdeanlewissoftwire commented on spec/features/schemes_spec.rb:
```diff
@@ -779,7 +779,7 @@
               it "allows to deactivate a location" do
                 click_link("Deactivate this location")
-                expect(page).to have_current_path("/schemes/#{scheme.id}/locations/#{location.id}/deactivate")
+                expect(page).to have_current_path("/schemes/#{scheme.id}/locations/#{location.id}/new-deactivation")
```
That is a bit nicer, but I think `new-deactivation,` `deactivate-confirm` and `deactivate` are ok for MVP
